File: csharp/2013-Detect-Squares.cs
public class DetectSquares {

  private Dictionary<(int x, int y), int> _pointsCounter = new();
    private List<(int x, int y)> _points = new();

    public DetectSquares() { }
    
    public void Add(int[] point) {
        var p = (point[0], point[1]);
        _points.Add(p);
        _pointsCounter[p] = 1 + _pointsCounter.GetValueOrDefault(p, 0);
    }
    
    public int Count(int[] point) {
        int px = point[0], py = point[1];
        int result = 0;

        foreach (var (x, y) in _points) {
            
            if (Math.Abs(px - x) != Math.Abs(py - y) 
                || x == px || y == py) { 
                continue;
            }
            result += _pointsCounter.GetValueOrDefault((px, y), 0) * _pointsCounter.GetValueOrDefault((x, py), 0);
        }
        return result;
    }
}

/**
 * Your DetectSquares object will be instantiated and called as such:
 * DetectSquares obj = new DetectSquares();
 * obj.Add(point);
 * int param_2 = obj.Count(point);
 */

File: ruby/2013-detect-squares.rb
class DetectSquares
    def initialize()
        @points_count = Hash.new(0)
        @points = []   
    end


    def add(point)
        @points_count[point] +=1
        @points.append(point)
    end

    def count(point)
        result = 0
        px = point[0]
        py = point[1]

        @points.each do |p|
            next if (((py-p[1]).abs != (px-p[0]).abs) || (p[0]==px) ||(p[1]==py))
            result += @points_count[[p[0],py]] * @points_count[[px,p[1]]]
        end

        return result
        
    end


end